wolfcrypt/src/ascon.c

```diff
@@ -0,0 +1,528 @@
+    (a)->s64[0] ^= (a)->s64[4];                                                \
+    (a)->s64[4] ^= (a)->s64[3];                                                \
+    (a)->s64[2] ^= (a)->s64[1];                                                \
+    tmp.s64[0] = (a)->s64[0] ^ (~(a)->s64[1] & (a)->s64[2]);                   \
+    tmp.s64[2] = (a)->s64[2] ^ (~(a)->s64[3] & (a)->s64[4]);                   \
+    tmp.s64[4] = (a)->s64[4] ^ (~(a)->s64[0] & (a)->s64[1]);                   \
+    tmp.s64[1] = (a)->s64[1] ^ (~(a)->s64[2] & (a)->s64[3]);                   \
+    tmp.s64[3] = (a)->s64[3] ^ (~(a)->s64[4] & (a)->s64[0]);                   \
+    tmp.s64[1] ^= tmp.s64[0];                                                  \
+    tmp.s64[3] ^= tmp.s64[2];                                                  \
+    tmp.s64[0] ^= tmp.s64[4];                                                  \
+    tmp.s64[2] = ~tmp.s64[2];                                                  \
+    /* 2.6.3 Linear Diffusion Layer */                                         \
+    (a)->s64[4] =                                                              \
+       tmp.s64[4] ^ rotrFixed64(tmp.s64[4],  7) ^ rotrFixed64(tmp.s64[4], 41); \
+    (a)->s64[1] =                                                              \
+       tmp.s64[1] ^ rotrFixed64(tmp.s64[1], 61) ^ rotrFixed64(tmp.s64[1], 39); \
+    (a)->s64[3] =                                                              \
+       tmp.s64[3] ^ rotrFixed64(tmp.s64[3], 10) ^ rotrFixed64(tmp.s64[3], 17); \
+    (a)->s64[0] =                                                              \
+       tmp.s64[0] ^ rotrFixed64(tmp.s64[0], 19) ^ rotrFixed64(tmp.s64[0], 28); \
+    (a)->s64[2] =                                                              \
+       tmp.s64[2] ^ rotrFixed64(tmp.s64[2],  1) ^ rotrFixed64(tmp.s64[2],  6); \
+    } while (0)
+
+#define p6(a) \
+    p(a, 0x96); \
+    p(a, 0x87); \
+    p(a, 0x78); \
+    p(a, 0x69); \
+    p(a, 0x5a); \
+    p(a, 0x4b)
+
+#define p8(a) \
+    p(a, 0xb4); \
+    p(a, 0xa5); \
+    p6(a)
+
+#define p12(a) \
+    p(a, 0xf0); \
+    p(a, 0xe1); \
+    p(a, 0xd2); \
+    p(a, 0xc3); \
+    p8(a)
+
+/* Needed layer to evaluate the macro values */
+#define _permutation(a, rounds) \
+    p ## rounds(a)
+
+#define permutation(a, rounds) \
+    _permutation(a, rounds)
+
+#endif
+
+/* AsconHash API */
+
+wc_AsconHash256* wc_AsconHash256_New(void)
+{
+    wc_AsconHash256* ret = (wc_AsconHash256*)XMALLOC(sizeof(wc_AsconHash256),
+            NULL, DYNAMIC_TYPE_ASCON);
+    if (ret != NULL) {
+        if (wc_AsconHash256_Init(ret) != 0) {
+            wc_AsconHash256_Free(ret);
+            ret = NULL;
+        }
+    }
+    return ret;
+}
+
+void wc_AsconHash256_Free(wc_AsconHash256* a)
+{
+    if (a != NULL) {
+        wc_AsconHash256_Deinit(a);
+        XFREE(a, NULL, DYNAMIC_TYPE_ASCON);
+    }
+}
+
+int wc_AsconHash256_Init(wc_AsconHash256* a)
```

Comment:
Should this have heap hint and devId for future API planning or just add an ex version when that time comes? Same for `wc_AsconHash256_New`

Reply:
I think those options fit more into an `_ex` version of the API. This one should be the simplest case that should cover most use cases.